Architecture Diagrams (#201)
mulesoft/docs-gateway: gateway/1.4/modules/ROOT/pages/flex-architecture-uc1.adoc

> = Use Case 1: Organization-Owned API Exposed to Organization-Owned API Consumer

Reviewer: ```suggestion
= Use Case 1: Organization-Owned API Exposed to an Organization-Owned API Consumer
```

Reviewer: Can we say "Exposed to an Organization Internal Consumer"? This sounds funny without the "an." Global for all use cases and pattern descriptions.

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

> Use case one covers all patterns of a trusted organization API exposed to a trusted organization API consumer.

Reviewer: ```suggestion
Use case one covers all patterns of a trusted organization's API exposed to a trusted organization's API consumer.
```

== Pattern 1: Organization Internal API Exposed to Organization Internal Consumer

Pattern examples:

> * Sales accessing a customer service API where both sales and customer service are internal areas of the customer organization

Reviewer: Sales application? Sales API?

> * Call center application calls API exposed by customer domain to retrieve customer information

Reviewer: ```suggestion
* Call center application calls the API exposed by the customer domain to retrieve customer information
```

> * Mobile bank mobile calls customer and account services to display account overview screen to end consumer

Reviewer: ```suggestion
* Mobile bank calls customer and account services to display the account overview screen to the end consumer
```

== Pattern 1A: Organization Internal API Exposed to Organization Internal Consumer Using the Same Vpc or Network

> In this use case pattern both services are available on the same infrastructure, for example, the same public cloud VPC, the same datacenter, or the same Kubernetes cluster. You can deploy Flex gateway in either sidecar mode or standalone gateway mode.

Reviewer: ```suggestion
In this use case pattern, both services are available on the same infrastructure, such as the same public cloud VPC, datacenter, or Kubernetes cluster. You can deploy Flex gateway in either sidecar mode or standalone gateway mode.
```

> image:ad-uc-1-p-1a.png[For both standalone and sidecar mode Flex Gateway and the APIs are in the same internal network.]

Reviewer: ```suggestion
image:ad-uc-1-p-1a.png[For both standalone and sidecar mode, Flex Gateway and the APIs are in the same internal network.]
```

> The following diagram demonstrates the physical implementation of a standalone Flex Gateway running in Connected Mode. A network policy (Netpol) permits ingress access only from Flex Gateway. You may configure the consumer application in the same or a different namespace from the Flex Gateway and the provider services.

Reviewer: ```suggestion
The following diagram shows the physical implementation of a standalone Flex Gateway running in Connected Mode. A network policy (Netpol) permits ingress access only from Flex Gateway. You can configure the consumer application in the same or different namespace from Flex Gateway and the provider services.
```

image:ad-uc-1-p-1a-physical.png[A more detailed view of UC1 contains load balancers and Netpol protection]

> When the API provider is located in a different kubernetes cluster from Flex Gateway and the consumer application, you can either deploy a Flex Gateway runtime on each of the clusters, as in Pattern 1B, or establish network restrictions in the secure channel between clusters, using Flex Gateway in only one of them as in the following diagram. The second cluster only allows connections from Flex Gateway and uses a standard ingress.

Reviewer: ```suggestion
When the API provider is located in a different Kubernetes cluster from Flex Gateway and the consumer application, you do either of the following:

* Deploy a Flex Gateway runtime on each of the clusters, as in Pattern 1B.
* Establish network restrictions in the secure channel between clusters, and establish Flex Gateway in only one of them, as shown in the following diagram. The second cluster allows connections only from Flex Gateway, and it uses a standard ingress.
```

> image:ad-uc-1-p-1b-separate-clusters.png[A detailed view of pattern 1A containing the necessary services to support Flex gateway] 

Reviewer: ```suggestion
image:ad-uc-1-p-1b-separate-clusters.png[A detailed view of pattern 1A, which contains the necessary services to support Flex gateway] 
```

> == Pattern 1B: Organization Internal API Exposed to Organization Internal Consumer Using a Different Vpc or Network

Reviewer: ```suggestion
== Pattern 1B: Organization Internal API Exposed to an Organization Internal Consumer that Uses a Different VPC or Network
```

> In this use case pattern, the API provider is in infrastructure isolated from the API consumer. However, there is a secure communication channel available between the two networks.

Reviewer: Can you simplify by saying "is isolated from the API consumer"?

> You deploy Flex Gateway as an ingress controller because traffic is considered external to the location of the provider. Flex Gateway receives and processes the incoming traffic sitting behind the firewall.

Reviewer: ```suggestion
You deploy Flex Gateway as an ingress controller because traffic is considered external to the location of the provider. Flex Gateway receives and processes the incoming traffic that sits behind the firewall.
```

You can use the Flex gateway to terminate TLS or mTLS connection if required.

> image:ad-uc-1-p-1b.png[Information travels from a trusted network to the internal network containing Flex Gateway]

Reviewer: ```suggestion
image:ad-uc-1-p-1b.png[Information travels from a trusted network to the internal network that contains Flex Gateway]
```

The following diagram demonstrates the physical implementation on Kubernetes.

> image:ad-uc-1-p-1b-physical.png[The previous networks are now kubernetes clusters and there are load balancers separating the APIs from Flex Gateway.] 

Reviewer: ```suggestion
image:ad-uc-1-p-1b-physical.png[The previous networks are now Kubernetes clusters and there are load balancers that separate the APIs from Flex Gateway.] 
```

== Pattern 2: Organization Internal API exposed Organization External Consumer

> In pattern two, the organization-owned API is in infrastructure isolated from the organization-owned API consumer, for example, a package delivery tracking API that is accessed from a user interface. The consumer can only access the exposed API via the internet. In this use case, you deploy Flex Gateway as an ingress controller. 

Reviewer: ```suggestion
In pattern two, the organization-owned API is in infrastructure isolated from the organization-owned API consumer; for example, in a package delivery tracking API that is accessed from a user interface. The consumer can access the exposed API only via the internet. In this use case, you deploy Flex Gateway as an ingress controller. 
```

image:ad-uc-1-p-2.png[Consumer application requests pass through the internet and a firewall before reaching Flex Gateway]

This example is similar to Pattern 1B. However, traffic arrives from the public internet instead of a private channel. Because traffic is coming from the public internet, you must add additional controls to strengthen communication security before reaching Flex Gateway.

> The following diagram demonstrates the physical implementation of organization internal API exposed to organization internal consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming external mTLS connection.

Reviewer: ```suggestion
The following diagram demonstrates the physical implementation of an organization's internal API exposed to an organization's internal consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming external mTLS connection.
```

image:ad-uc-1-p-2-physical.png[The previous networks are now Kubernetes clusters, there are load balancers separating the APIs from Flex Gateway, and mTLS connections to external networks.]

== Pattern 3: Organization External API Exposed to Organization Internal Consumer

> Pattern three is the inverse of pattern two. You can use Flex Gateway to monitor, control, and enrich the flows to external services, for example by rate limiting the external access or metrics and analytics tracking for this external service connection.

Reviewer: ```suggestion
Pattern three is the inverse of pattern two. You can use Flex Gateway to monitor, control, and enrich the flows to external services. For example, you can use Flex Gateway to rate limit external access, or you can use to track metrics and analytics for this external service connection.
```

Reviewer: The sentence above was confusing, so I proposed splitting into two sentences. I am not sure if I interpreted your sentence correctly.

Pattern examples:

> * Payments service needs to obtain exchange rates from an external service provider.

Reviewer: For consistency with the next bullet - "A payments service..."

> * Payments service needs to obtain exchange rates from an external service provider.

Reviewer: ```suggestion
* A payments service must obtain exchange rates from an external service provider.
```

* An internal API Consumer accesses the organization’s external ServiceNow instance (SaaS service).

> For pattern three, you deploy Flex Gateway as a standalone gateway acting as an egress container.

Reviewer: ```suggestion
For pattern three, you deploy Flex Gateway as a standalone gateway that acts as an egress container.
```

image:ad-uc-1-p-3.png[Flex Gateway is deployed as an egress controller to monitor incoming traffic from the external API.]

> The following diagram demonstrates the physical implementation of organization external API exposed to an organization internal consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming external mTLS connection.

Reviewer: ```suggestion
The following diagram demonstrates the physical implementation of an organization's external API exposed to an organization's internal consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming external mTLS connection.
```

> image:ad-uc-1-p-3-physical.png[The previous networks are now kubernetes clusters, there are load balancers separating the APIs from Flex Gateway, and mTLS connections to external networks.]

Reviewer: Global - initial cap Kubernates
```suggestion
image:ad-uc-1-p-3-physical.png[The previous networks are now kubernetes clusters. There are load balancers that separate the APIs from Flex Gateway and mTLS connections to external networks.]
```

== Pattern 4: Organization External API Exposed to Organization External Consumer

> In pattern four, Flex Gateway acts as an intermediary between two organization external entities, for example, an organization’s WorkDay instance (SaaS service) accesses the organization’s ServiceNow instance (SaaS service). Flex Gateway runs as an Ingress and Egress controller applying secured communication and policy enforcement.

Reviewer: ```suggestion
In pattern four, Flex Gateway acts as an intermediary between two organizations' external entities, for example, when an organization’s WorkDay instance (SaaS service) accesses the organization’s ServiceNow instance (SaaS service). In this pattern, Flex Gateway runs as an Ingress and Egress controller that applies secured communication and policy enforcement.
```

image:ad-uc-1-p-4.png[Flex Gateway acts as an intermediary between an external application and external API.]

> The following diagram demonstrates the physical implementation of an external organization API exposed to an external organization Consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming Org External mTLS connection.

Reviewer: ```suggestion
The following diagram demonstrates the physical implementation of an external organization's API exposed to an external organization's Consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming Org External mTLS connection.
```

> image:ad-uc-1-p-4-physical.png[A detailed view of pattern four containing the necessary services to support Flex gateway]

Reviewer: ```suggestion
image:ad-uc-1-p-4-physical.png[A detailed view of pattern four, which contains the necessary services to support Flex gateway]
```

= Use Case 1: Organization-Owned API Exposed to an Organization-Owned API Consumer
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

> Use case one covers all patterns of a trusted organization's API exposed to a trusted organization's API consumer.

Reviewer: ```suggestion
Use case one covers all patterns of a scenario in which a trusted organization's API is exposed to a trusted organization's API consumer.
```

== Pattern 1: Organization Internal API Exposed to an Organization Internal Consumer

Pattern examples:

> * A sales application accessing a customer service API where both sales and customer service are internal areas of the customer organization

Reviewer: ```suggestion
* A sales application accesses a customer service API in which both sales and customer service are internal areas of the customer organization
```

* A call center application calls the API exposed by the customer domain to retrieve customer information
* A mobile bank calls customer and account services to display the account overview screen to the end consumer

> === Pattern 1A: Organization Internal API Exposed to an Organization Internal Consumer Using the Same Vpc or Network

Reviewer: ```suggestion
=== Pattern 1A: Organization Internal API Exposed to an Organization Internal Consumer that Uses the Same VPC or Network
```

> In this use case pattern, both services are available on the same infrastructure, such as the same public cloud VPC, datacenter, or Kubernetes cluster. You can deploy Flex Gateway as either a sidecar or as a standalone gateway.

Reviewer: ```suggestion
In this use case pattern, both services are available on the same infrastructure, such as the same public cloud VPC, datacenter, or Kubernetes cluster. You can deploy Flex Gateway as either a sidecar or a standalone gateway.
```

image:ad-uc-1-p-1a.png[For both standalone and sidecar mode, Flex Gateway and the APIs are in the same internal network.]

The following diagram shows the physical implementation of a standalone Flex Gateway running in Connected Mode. A network policy (Netpol) permits ingress access only from Flex Gateway. You can configure the consumer application in the same or different namespace from Flex Gateway and the provider services.

image:ad-uc-1-p-1a-physical.png[A more detailed view of UC1 contains load balancers and Netpol protection]

> When the API provider is located in a different Kubernetes cluster from Flex Gateway and the consumer application, you can so either of the following:

Reviewer: ```suggestion
When the API provider is located in a different Kubernetes cluster from Flex Gateway and the consumer application, you can do either of the following:
```

* Deploy a Flex Gateway runtime on each of the clusters, as in Pattern 1B.
* Establish network restrictions in the secure channel between clusters and deploy Flex Gateway in only one of them, as shown in the following diagram. The second cluster only allows connections from a Flex Gateway that acts as an ingress controller.

image:ad-uc-1-p-1b-separate-clusters.png[A detailed view of pattern 1A, which contains the necessary services to support Flex gateway] 

> === Pattern 1B: Organization Internal API Exposed to an Organization Internal Consumer Using a Different Vpc or Network

Reviewer: ```suggestion
=== Pattern 1B: Organization Internal API Exposed to an Organization Internal Consumer that Uses a Different VPC or Network
```

> In this use case pattern, the API provider is isolated from the API consumer. However, there is a secure communication channel available between the two networks.

Reviewer: ```suggestion
In this pattern, the API provider is isolated from the API consumer. However, there is a secure communication channel available between the two networks.
```

You deploy Flex Gateway as an ingress controller behind a firewall because traffic is considered external to the location of the provider. Flex Gateway receives and processes the incoming traffic.

> You can use the Flex gateway to terminate TLS or mTLS connection if required.

Reviewer: ```suggestion
You can use Flex Gateway to terminate the TLS or mTLS connection if required.
```

image:ad-uc-1-p-1b.png[Information travels from a trusted network to the internal network that contains Flex Gateway.]

The following diagram shows the physical implementation on Kubernetes.

image:ad-uc-1-p-1b-physical.png[The previous networks are now Kubernetes clusters and there are load balancers that separate the APIs from Flex Gateway.] 

== Pattern 2: Organization Internal API exposed an Organization External Consumer
In pattern two, the organization-owned API is isolated from the organization-owned API consumer; for example, in a package delivery tracking API that is accessed from a user interface. The consumer can access the exposed API only via the internet. In this use case, you deploy Flex Gateway as an ingress controller. 

image:ad-uc-1-p-2.png[Consumer application requests pass through the internet and a firewall before reaching Flex Gateway.]

> This example is similar to Pattern 1B. However, traffic arrives from the public internet instead of a private channel. Because traffic is coming from the public internet, you must add additional controls to strengthen communication security before reaching Flex Gateway.

Reviewer: "This pattern" instead of "This example"?

> This example is similar to Pattern 1B. However, traffic arrives from the public internet instead of a private channel. Because traffic is coming from the public internet, you must add additional controls to strengthen communication security before reaching Flex Gateway.

Reviewer: ```suggestion
This pattern is similar to Pattern 1B. However, in this pattern, traffic arrives from the public internet instead of from a private channel. Because traffic is coming from the public internet, you must add additional controls to strengthen communication security before reaching Flex Gateway.
```

Reviewer: "...strengthen communication security" of what?

Maybe: 
...strengthen the security of communication channels that lead to Flex Gateway"

> The following diagram shows the physical implementation of an organization's internal API exposed to an organization's internal consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming external mTLS connection.

Reviewer: ```suggestion
The following diagram shows the physical implementation of a scenario in which an organization's internal API is exposed to an organization's internal consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming external mTLS connection.
```

image:ad-uc-1-p-2-physical.png[The previous networks are now Kubernetes clusters, there are load balancers separating the APIs from Flex Gateway, and mTLS connections to external networks.]

== Pattern 3: Organization External API Exposed to an Organization Internal Consumer

> In pattern three, the API Consumer is in the same network as Flex Gateway. You can use Flex Gateway to monitor, control, and enrich the flows to external services, for example Flex Gateway can provide rate limiting for the external access traffic or metrics and analytics tracking for this external service connection.

Reviewer: ```suggestion
In pattern three, the API Consumer is in the same network as Flex Gateway. You can use Flex Gateway to monitor, control, and enrich the flows to external services. For example, Flex Gateway can provide rate limiting for the external access traffic, or metrics and analytics tracking for this external service connection.
```

Pattern examples:

* A payments service must obtain exchange rates from an external service provider.
* An internal API Consumer accesses the organization’s external ServiceNow instance (SaaS service).

For pattern three, you deploy Flex Gateway as a standalone gateway that acts as an egress container.

image:ad-uc-1-p-3.png[Flex Gateway is deployed as an egress controller to monitor incoming traffic from the external API.]

> The following diagram shows the physical implementation of an organization's external API exposed to an organization's internal consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming external mTLS connection.

Reviewer: ```suggestion
The following diagram shows the physical implementation of a scenario in which an organization's external API is exposed to an organization's internal consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming external mTLS connection.
```

image:ad-uc-1-p-3-physical.png[The previous networks are now kubernetes clusters. There are load balancers that separate the APIs from Flex Gateway and mTLS connections to external networks.]

== Pattern 4: Organization External API Exposed to an Organization External Consumer

> In pattern four, Flex Gateway acts as an intermediary between two organizations' external entities, for example, when an organization’s WorkDay instance (SaaS service) accesses the organization’s ServiceNow instance (SaaS service). In this pattern, Flex Gateway runs as an Ingress and Egress controller that applies secured communication and policy enforcement.

Reviewer: ```suggestion
In pattern four, Flex Gateway acts as an intermediary between two organizations' external entities, for example, when an organization’s WorkDay instance (SaaS service) accesses the organization’s ServiceNow instance (SaaS service). In this pattern, Flex Gateway runs as an ingress and egress controller that applies secured communication and policy enforcement.
```

image:ad-uc-1-p-4.png[Flex Gateway acts as an intermediary between an external application and external API.]

> The following diagram shows the physical implementation of an external organization's API exposed to an external organization's Consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming organization external mTLS connection.

Reviewer: Make Consumer lower-case?

> The following diagram shows the physical implementation of an external organization's API exposed to an external organization's Consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming organization external mTLS connection.

Reviewer: ```suggestion
The following diagram shows the physical implementation of an external organization's API exposed to an external organization's Consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming organization's external mTLS connection.
```

> The following diagram shows the physical implementation of an external organization's API exposed to an external organization's Consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming organization external mTLS connection.

Reviewer: ```suggestion
The following diagram shows the physical implementation of a scenario in which an external organization's API is exposed to an external organization's Consumer on Kubernetes. The diagram assumes that the firewall terminates the incoming organization's external mTLS connection.
```
